### ISSUE TITLE
ENH: pass options from scipy.ndimage to our wrappers

### ISSUE THREAD

**Issue by sciunto** [scikit-image/scikit-image#2597]
Part of #2510

In skimage/morphology/binary.py
at least, `binary_erosion` `binary_dilation` can get new arguments as ndimage supports some nice features like the kwarg `iterations` See https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.binary_dilation.html#scipy.ndimage.binary_dilation

Also, add in the documentation a note saying that our functions are based on ndimage.

May have similar features in In skimage/morphology/grey.py...

**Comment by soupault** [comment]
xref https://github.com/scikit-image/scikit-image/issues/1917

**Comment by nrweir** [comment]
I use these binary morphology functions often in my research (both the scipy versions and the skimage wrapper) and would be interested in making these changes, if desired. I've read the contributing.txt and will do my best to fulfill requirements in my version (this is my first real open-source contribution and may require some patience as I get used to coding standards)

**Comment by stefanv** [comment]
Thanks, @nrweir, looking forward to your first PR!

**Comment by nrweir** [comment]
TL;DR previous versions of skimage `black_tophat` gives different output than the current `ndi.morphology.black_tophat` due how it handles edges, and I don't know which you would prefer be implemented in the new version of scikit-image's `black_tophat`.

In the past, scikit-image `morphology.grey.black_tophat` used `morphology.grey.closing`, which automatically pads image edges when using structuring elements with even length axes. This increases the size of edge regions, which changes how they're processed by filters which use `morphology.grey.closing` (namely, `black_tophat`). This got caught in Travis CI testing, where my new version of `grey.black_tophat` (which utilized `ndi.morphology.black_tophat`) failed a couple of the tests (see output attached) because the test's expected output requires the image to be padded, which `ndi.morphology.black_tophat` doesn't do. I'm attaching a pyplot showing the original test image, the `ndi.morphology.black_tophat` result of processing it,  the padded version, and the result that previous versions of scikit-image `black_tophat` would have provided.

_my personal opinion (feel free to ignore)_: The new arguments to `ndi.morphology.black_tophat` that I'm implementing provide several options for how to handle edges: 'reflect', 'constant', 'nearest', 'mirror', and 'wrap'. I could add one more option to that list, 'pad', which would implement padding as was implemented previously in scikit-image. If you like, I could make that the default to maintain consistent output with previous versions of scikit-image. Alternatively, the expected test output could be updated to reflect `ndi.morphology.black_tophat` outputs and documentation could be updated to indicate that the new version of scikit-image handles image edges differently for the filter, and to use `mode='pad'` to produce the output from previous versions.

Sorry about the long comment, trying to cover all the bases.

version of grey.py used here: https://github.com/nrweir/scikit-image/blob/scipy-binary-args/skimage/morphology/grey.py

Travis CI output
----------------

=================================== FAILURES ===================================
_____________________ TestMorphology.test_gray_morphology ______________________
self = <skimage.morphology.tests.test_grey.TestMorphology testMethod=test_gray_morphology>
    def test_gray_morphology(self):
        expected = dict(np.load(
            os.path.join(data_dir, 'gray_morph_output.npz')))
        calculated = self._build_expected_output()
       assert_equal(expected, calculated)
E       AssertionError: 
E       Arrays are not equal
E       key='square_2_black_tophat'
E       
E       (mismatch 5.0%)
E        x: array([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  6,  0,  0,  0,  2,  0,
E                0,  0,  0,  1,  0,  0,  0,  0,  0,  0,  0,  0,  0],
E              [ 0,  0,  0,  2,  1,  0,  0,  0,  0,  8,  8,  0,  0,  0,  0,  0,  0,...
E        y: array([[  7,   0,   0,   0,   0,  23,  47,  32,   1,   0,   4,  79, 117,
E               107, 102, 113,  95,  44,   0,   2,   6,   0,   0,   4,   2,   0,
E                 0,   0,   0,   0],...
skimage/morphology/tests/test_grey.py:49: AssertionError

Sample processing output showing difference
---------------------------------------------

![image](https://user-images.githubusercontent.com/13022883/33676694-74cd30e0-da84-11e7-8f54-d6b79459a6e9.png)

(note that this only shows the top middle of the padded image due to dimensions change)

**Comment by soupault** [comment]
My suggestion would be to align with `scipy` defaults as much as possible. This would help us to reduce the maintenance burden for `morphology` functions (not just functionality update, but also testing).
@nrweir Do you have an estimate on how many of `scikit-image` features depend on internally shipped morphology functions implementations?

**Comment by nrweir** [comment]
I'll have a look and get back to you today or tomorrow. I'll also double-check the other binary and grey morphology functions to make sure they don't produce outputs inconsistent with their `scipy` equivalents.

**Comment by nrweir** [comment]
@soupault Very, very few features import and use `scikit-image`'s morphology functions - in combing through I only found one case (`segmentation.boundaries.find_boundaries`), which uses `morphology.grey.erosion` and `morphology.grey.dilation`. It's possible I missed some though.

Makes sense to me to align with `scipy` defaults, I'll implement that for `morphology.grey.black_tophat`. This would also require making some changes to the existing tests to reflect the new outputs for builds to pass.
